doc/developer/design/20231113_optimizer_notice_catalog.md

```diff
@@ -0,0 +1,480 @@
+  provide more context, a minimal reproducing example, and a corrective action
+  for the example for each notice type for customers that want to understand
+  where and why a specific notice is emitted.
+
+### Catalog Schema
+
+#### <a id="mz_optimizer_notices"></a> `mz_internal.mz_optimizer_notices` (new)
+
+Optimizer notices are exposed through the `mz_internal.mz_optimizer_notices`
+view with the following columns.
+
+| Field                   | Type                         | Meaning                                                                                                                                           |
+| ----------------------- | ---------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `notice_type`           | `text`                       | The notice type. Each type will correspond to a Materialize docs sub-page.                                                                        |
+| `message`               | `text`                       | A brief description of the issue highlighted by this notice.                                                                                      |
+| `hint`                  | `text`                       | A high-level hint that tells the user what can be improved.                                                                                       |
+| `action`                | `text`                       | A concrete action that will resolve the notice.                                                                                                   |
+| `redacted_message`      | `text`                       | A redacted version of the `message` column. `NULL` if no redaction is needed.                                                                     |
+| `redacted_hint`         | `text`                       | A redacted version of the `hint` column. `NULL` if no redaction is needed.                                                                        |
+| `redacted_action`       | `text`                       | A redacted version of the `action` column. `NULL` if no redaction is needed.                                                                      |
+| `action_type`           | `text`                       | The type of the `action` string (`sql_statements` for a valid SQL string or `plain_text` for plain text).                                         |
+| `object_id`             | `text`                       | The ID of the materialized view or index. Corresponds to [`mz_objects.id`](../mz_catalog/#mz_objects). For global notices, this column is `NULL`. |
+| `dependency_ids`        | `text list`                  | A list of dependency IDs that need to exist for this notice to be still valid. Corresponds to [`mz_objects.id`](../mz_catalog/#mz_objects).       |
```

Comment:
I'm thinking about the exact connection between `object_id` and `dependency_ids`. For an example, let's consider the `CommonArrangedInput` notice (I just made up this name; it's https://github.com/MaterializeInc/materialize/issues/20802).

In this example, would `dependency_ids` list all those objects that arrange the same one input? But then if one of these objects is dropped, the notice might still be valid if originally there were more than two objects arranging the same input. It becomes invalid only if all but one of the ids in `dependency_ids` is dropped.

Another question for this example: what will be the object id? Will we have one row for each object in `dependency_ids`?

Reply:
Maybe another way to model this is to have `dependant_ids` (as a separate relation or a list) instead of `object_id`. This will contain one entry for notices scoped to a single element and multiple entries for things like `CommonArrangedInput`.

Reply:
Hmm, ok, so this will be used by Rust code when computing table updates, and not by some SQL in a view that just simply hides the row when any (or all) of the things in `dependency_ids` is gone, right?

Reply:
> Maybe another way to model this is to have dependant_ids (as a separate relation or a list) instead of object_id. This will contain one entry for notices scoped to a single element and multiple entries for things like CommonArrangedInput.

Sounds good.

Reply:
Yes, sadly I don't see a good way around that. The `BuiltinTable` API is written that way - it is meant to be a store for side-effects that reflects certain parts of the in-memory state of the system. The adapter code does not have access to the contents of any `BuiltinTable` - it can only emit a retraction (based on its current state) and create an insertion against table.

@ParkMyCar please disagree with me here if this is not correct, but this was my impression from reading the code.

Reply:
Personally I would love if we could unify the two, specifically if we could also eagerly retract from `mz_all_optimizer_notices` for the `dependency_ids` column.